### PR TITLE
tests added (ViewRenderer#extractControllerName)

### PR DESCRIPTION
**Pull request by kamarton**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #61

I seems the `ViewRenderer#extractControllerName` is not working fine.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 ## 5.0.1 under development
 
-- no changes in this release.
+- Bug #61: Fixed incorrect receipt a controller name based on controller instance (@vjik, @kamarton)
```
Comment (samdark) [inline CHANGELOG.md:5]:
```suggestion
- Bug #61: Fixed getting incorrect controller name based on controller instance (@vjik, @kamarton)
```

Comment (vjik) [reply CHANGELOG.md]: done

File: src/ViewRenderer.php [hunk 2]
```diff
@@ -561,7 +561,7 @@ private function extractControllerName(object $controller): string
             return $cache[$class];
         }
 
-        $regexp = '/((?<=controller\\\|s\\\)(?:[\w\\\]+)|(?:[a-z]+))controller/iuU';
+        $regexp = '/((?<=controller\\\|controllers\\\)(?:[\w\\\]+)|(?:[a-z]+))controller/iuU';
```
Comment (samdark) [inline src/ViewRenderer.php:564]:
```suggestion
        $regexp = '/((?<=controllers?\\\)(?:[\w\\\]+)|(?:[a-z]+))controller/iuU';
```

Comment (vjik) [reply src/ViewRenderer.php:564]: It's not work: `preg_match(): Compilation failed: lookbehind assertion is not fixed length at offset 1`

Comment (kamarton) [reply src/ViewRenderer.php:564]: I suggests:
https://regex101.com/r/5M2D54/3
```php
$regexp = '!(?:controllers?\\(?<subpath>[^\\]+(?:\\[^\\]+)*\\))?(?<path>(?:[^\\]+))controller$!iu';
if(preg_match($regexp, $class, $m) !== 1) {
  throw ...
}
$name=($m['subpath'] ?? '') . $m['path'];
```

File: tests/ViewRendererTest.php [hunk 3]
```diff
@@ -175,7 +179,44 @@ public function testRenderWithLocale(): void
...
+                "/sub-namespace/sub-namespace2/fake"
+            ],
+            "controller-name-and-controllers-namespace-with-two-subnamespaces" => [
+                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
+                "/sub-namespace/sub-namespace2/fake"
+            ],
+            "controller-name-with-multiple-subnamespaces" => [
+                new Support\NotCntrls\SubNamespace\FakeController(),
+                '/fake',
+            ],
```
Comment (arogachev) [inline tests/ViewRendererTest.php:205]:
```suggestion
            "controller-name" => [new Support\FakeController(), '/fake'],
            "controller-name-and-namespace" => [
                new Support\Controller\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            "controller-name-and-controllers-namespace" => [
                new Support\Controllers\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            "controller-name-and-namespace-with-two-subnamespaces" => [
                new Support\Controller\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            "controller-name-and-controllers-namespace-with-two-subnamespaces" => [
                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            "controller-name-with-multiple-subnamespaces" => [
                new Support\NotCntrls\SubNamespace\FakeController(),
                '/fake',
            ],
```

File: src/ViewRenderer.php (hunk 2 above)
Comment (arogachev) [reply src/ViewRenderer.php:564]: Unfortunately the last snippet doesn't work too. Gives the following error:

```
preg_match(): Compilation failed: missing terminating ] for character class at offset 76
```

Also it's more complicated and harder to read / maintain in my opinion. With the amount of variants and their length ("controller" and "controller**s**"), duplicating the whole world with different ending is acceptable and a better option I think.

File: tests/ViewRendererTest.php [hunk 4]
```diff
@@ -175,7 +179,44 @@ public function testRenderWithLocale(): void
...
+                '/sub-namespace/sub-namespace2/fake',
+            ],
+            "controller-name-and-controllers-namespace-with-two-subnamespaces" => [
+                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
+                '/sub-namespace/sub-namespace2/fake',
+            ],
+            "controller-name-with-multiple-subnamespaces" => [
+                new Support\NotCntrls\SubNamespace\FakeController(),
+                '/fake',
+            ],
```
Comment (arogachev) [inline tests/ViewRendererTest.php:205]:
```suggestion
            'controller-name' => [new Support\FakeController(), '/fake'],
            'controller-name-and-namespace' => [
                new Support\Controller\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            'controller-name-and-controllers-namespace' => [
                new Support\Controllers\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            'controller-name-and-namespace-with-two-subnamespaces' => [
                new Support\Controller\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            'controller-name-and-controllers-namespace-with-two-subnamespaces' => [
                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            'controller-name-with-multiple-subnamespaces' => [
                new Support\NotCntrls\SubNamespace\FakeController(),
                '/fake',
            ],
```

File: tests/ViewRendererTest.php [hunk 5]
```diff
@@ -175,7 +179,44 @@ public function testRenderWithLocale(): void
...
+                '/sub-namespace/sub-namespace2/fake',
+            ],
+            'controller-name-and-controllers-namespace-with-two-subnamespaces' => [
+                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
+                '/sub-namespace/sub-namespace2/fake',
+            ],
+            'controller-name-with-multiple-subnamespaces' => [
+                new Support\NotCntrls\SubNamespace\FakeController(),
+                '/fake',
+            ],
```
Comment (arogachev) [inline tests/ViewRendererTest.php:205]:
```suggestion
            'controller name, no "controller" / "controllers" namespaces, no subnamespaces' => [new Support\FakeController(), '/fake'],
            'controller name, "controller" namespace, 1 subnamespace' => [
                new Support\Controller\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            'controller name, "controllers" namespace, 1 subnamespace' => [
                new Support\Controllers\SubNamespace\FakeController(),
                '/sub-namespace/fake',
            ],
            'controller name, "controller" namespace, 2 subnamespaces' => [
                new Support\Controller\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            ''controller name, "controllers" namespace, 2 subnamespaces' => [
                new Support\Controllers\SubNamespace\SubNamespace2\FakeController(),
                '/sub-namespace/sub-namespace2/fake',
            ],
            'controller name, without "controller" / "controllers" namespaces, subnamespaces' => [
                new Support\NotCntrls\SubNamespace\FakeController(),
                '/fake',
            ],
```

File: tests/ViewRendererTest.php [hunk 6]
```diff
@@ -175,7 +179,44 @@ public function testRenderWithLocale(): void
...
+            'controller name, "controllers" namespace, 1 subnamespace' => [
+                new Support\Controllers\SubNamespace\FakeController(),
+                '/sub-namespace/fake',
+            ],
+            'controller name, "controller" namespace, 2 subnamespaces' => [
+                new Support\Controller\SubNamespace\SubNamespace2\FakeController(),
+                '/sub-namespace/sub-namespace2/fake',
+            ],
+            ''controller name, "controllers" namespace, 2 subnamespaces' => [
```
Comment (arogachev) [inline tests/ViewRendererTest.php:198]:
```suggestion
            'controller name, "controllers" namespace, 2 subnamespaces' => [
```

**Review by arogachev — APPROVED** [review]

**Pull request merged**